reef: cephfs_mirror: use snapdiff api for incremental syncing

joscollin opened the pull request:
backport tracker: https://tracker.ceph.com/issues/65222

---

backport of https://github.com/ceph/ceph/pull/54633
parent tracker: https://tracker.ceph.com/issues/61334

this backport was staged using ceph-backport.sh version 16.0.0.6848
find the latest version at https://github.com/ceph/ceph/blob/main/src/script/ceph-backport.sh

lxbsz commented:
jenkins test make check

rishabh-d-dave commented:
This needs a re-run and a rebuild since binaries built on shaman have been deleted by now . See - https://tracker.ceph.com/issues/67493#note-11.

vshankar commented:
@joscollin PTAL at the failure - https://pulpito.ceph.com/rishabh-2024-09-10_14:18:13-fs-wip-rishabh-testing-20240910.073325-reef-distro-default-smithi/7899232/

(which is reproducible)

joscollin commented:
> @joscollin PTAL at the failure - https://pulpito.ceph.com/rishabh-2024-09-10_14:18:13-fs-wip-rishabh-testing-20240910.073325-reef-distro-default-smithi/7899232/
> 
> (which is reproducible)

@vshankar 
This is a known issue [1] https://tracker.ceph.com/issues/68567. As I remember, this failure exists even before `snapdiff` change and still exist in main. This is getting failed intermittently. So I did a check. My observation is: The source and destination snapshots doesn't show any differences and it's been synced correctly. So it's failing because of something else, different from incremental sync, which need to be RCA'ed.

My opinion is we shouldn't block this PR because of this and should be tracked separately [1].

joscollin commented:
> @joscollin PTAL at the failure - https://pulpito.ceph.com/rishabh-2024-09-10_14:18:13-fs-wip-rishabh-testing-20240910.073325-reef-distro-default-smithi/7899232/
> 
> (which is reproducible)

@vshankar 
Please include this fix https://github.com/ceph/ceph/pull/61068 and rerun the QA.

joscollin commented:
@vshankar Added more commits. Please re-run the QA.

vshankar commented:
jenkins retest this please

vshankar commented:
Blocked due to: https://github.com/ceph/ceph/pull/54633#pullrequestreview-2484004834

joscollin commented:
Fix created: https://github.com/ceph/ceph/pull/61539.

mchangir commented:
jenkins test docs

mchangir commented:
jenkins test make check

mchangir commented:
This PR is under test in [https://tracker.ceph.com/issues/69881](https://tracker.ceph.com/issues/69881).

joscollin commented:
> This PR is under test in https://tracker.ceph.com/issues/69881.

@mchangir This PR is `DNM` and waiting for the fix https://github.com/ceph/ceph/pull/61539.

mchangir commented:
> > This PR is under test in https://tracker.ceph.com/issues/69881.
> 
> @mchangir This PR is `DNM` and waiting for the fix #61539.

Thanks for the reminder.

joscollin commented:
jenkins test api

joscollin commented:
> @joscollin This backport should include commits from https://github.com/ceph/ceph/pull/62250/commits except the blockdiff parts. IOW, commits [d9ac431](https://github.com/ceph/ceph/commit/d9ac4315154a73c7bc84e94f55a9966c115adaca) and [ae9b035](https://github.com/ceph/ceph/commit/ae9b0350e64e43dfd612b71435c9d288d2669a1e)

@vshankar It's better don't backport `SyncMechanism` to reef, as there are several conflicts. Updated the PR.

vshankar commented:
> > @joscollin This backport should include commits from https://github.com/ceph/ceph/pull/62250/commits except the blockdiff parts. IOW, commits [d9ac431](https://github.com/ceph/ceph/commit/d9ac4315154a73c7bc84e94f55a9966c115adaca) and [ae9b035](https://github.com/ceph/ceph/commit/ae9b0350e64e43dfd612b71435c9d288d2669a1e)
> 
> @vshankar It's better don't backport `SyncMechanism` to reef, as there are several conflicts. Updated the PR.

My concern is that the changes in main (excluding the blockdiff integration) might be fixing some more bugs, which might be missed if we do not include those changes.

How bad are the conflicts?

vshankar commented:
> > > > @joscollin This backport should include commits from https://github.com/ceph/ceph/pull/62250/commits except the blockdiff parts. IOW, commits [d9ac431](https://github.com/ceph/ceph/commit/d9ac4315154a73c7bc84e94f55a9966c115adaca) and [ae9b035](https://github.com/ceph/ceph/commit/ae9b0350e64e43dfd612b71435c9d288d2669a1e)
> > > 
> > > 
> > > @vshankar It's better don't backport `SyncMechanism` to reef, as there are several conflicts. Updated the PR.
> > 
> > 
> > My concern is that the changes in main (excluding the blockdiff integration) might be fixing some more bugs, which might be missed if we do not include those changes.
> 
> As there are conflicts, I've added the fixes from #61539 to this PR, if that's what you meant^.

Yeh. But to be on the safer side it would be good to include the changes from main branch for cephfs-mirror.

> 
> > How bad are the conflicts?
> 
> One of them is really bad and needs a closer look what's caused it in reef.

I would suggest checking it if its feasible.

joscollin commented:
> > > > > @joscollin This backport should include commits from https://github.com/ceph/ceph/pull/62250/commits except the blockdiff parts. IOW, commits [d9ac431](https://github.com/ceph/ceph/commit/d9ac4315154a73c7bc84e94f55a9966c115adaca) and [ae9b035](https://github.com/ceph/ceph/commit/ae9b0350e64e43dfd612b71435c9d288d2669a1e)
> > > > 
> > > > 
> > > > @vshankar It's better don't backport `SyncMechanism` to reef, as there are several conflicts. Updated the PR.
> > > 
> > > 
> > > My concern is that the changes in main (excluding the blockdiff integration) might be fixing some more bugs, which might be missed if we do not include those changes.
> > 
> > 
> > As there are conflicts, I've added the fixes from #61539 to this PR, if that's what you meant^.
> 
> Yeh. But to be on the safer side it would be good to include the changes from main branch for cephfs-mirror.
> 
> > > How bad are the conflicts?
> > 
> > 
> > One of them is really bad and needs a closer look what's caused it in reef.
> 
> I would suggest checking it if its feasible.

ok, Let's do it afresh today. I'll pull the required commits from https://github.com/ceph/ceph/pull/62250 this time.

joscollin commented:
jenkins test docs

joscollin commented:
> > > > > @joscollin This backport should include commits from https://github.com/ceph/ceph/pull/62250/commits except the blockdiff parts. IOW, commits [d9ac431](https://github.com/ceph/ceph/commit/d9ac4315154a73c7bc84e94f55a9966c115adaca) and [ae9b035](https://github.com/ceph/ceph/commit/ae9b0350e64e43dfd612b71435c9d288d2669a1e)
> > > > 
> > > > 
> > > > @vshankar It's better don't backport `SyncMechanism` to reef, as there are several conflicts. Updated the PR.
> > > 
> > > 
> > > My concern is that the changes in main (excluding the blockdiff integration) might be fixing some more bugs, which might be missed if we do not include those changes.
> > 
> > 
> > As there are conflicts, I've added the fixes from #61539 to this PR, if that's what you meant^.
> 
> Yeh. But to be on the safer side it would be good to include the changes from main branch for cephfs-mirror.
> 
> > > How bad are the conflicts?
> > 
> > 
> > One of them is really bad and needs a closer look what's caused it in reef.
> 
> I would suggest checking it if its feasible.

@vshankar Done. Cherry-picked with minimum conflicts and resolved them.

joscollin commented:
jenkins test windows

joscollin commented:
jenkins test windows

vshankar commented:
> @vshankar This is ready for QA.

Brilliant! I will include it in my next backport run.

vshankar commented:
This PR is under test in [https://tracker.ceph.com/issues/71521](https://tracker.ceph.com/issues/71521).

vshankar commented:
@joscollin This requires an additional backport for commits from https://tracker.ceph.com/issues/70287

joscollin commented:
> @joscollin This requires an additional backport for commits from https://tracker.ceph.com/issues/70287

We have a cherry-pick conflict there. I'll work on it and update.

joscollin commented:
> @joscollin This requires an additional backport for commits from https://tracker.ceph.com/issues/70287

@vshankar Done. Conflicts are resolved.

vshankar commented:
> > @joscollin This requires an additional backport for commits from https://tracker.ceph.com/issues/70287
> 
> @vshankar Done. Conflicts are resolved.

@joscollin The commit message in the conflicting changes needs to include the "Conflicts section".

joscollin commented:
jenkins test api

vshankar commented:
> > > > @joscollin This requires an additional backport for commits from https://tracker.ceph.com/issues/70287
> > > 
> > > 
> > > @vshankar Done. Conflicts are resolved.
> > 
> > 
> > @joscollin The commit message in the conflicting changes needs to include the "Conflicts section".
> 
> @vshankar I've made it without conflicts by cherry-picking additional commits. It had conflicts, but the additional cherry-picking automatically resolved it.

OK. I was missing the piece of information as you mentioned in comment https://github.com/ceph/ceph/pull/58985#issuecomment-2940437284 that the cherry-pick had conflicts.

hkadam134 commented:
The PR is under test in - https://tracker.ceph.com/issues/72107

joscollin commented:
> > https://pulpito.ceph.com/hyelloji-2025-07-15_05:38:37-fs:mirror-wip-hemanth1-testing-2025-07-14-1140-reef-distro-default-smithi/8387974/
> > @vshankar Incremental sync fails with the reef backport. Putting DNM.
> 
> Please debug the failure @joscollin

@vshankar 
There are 2 sync failures seen in the logs. Looks like the reported one is readdir_snapdiff


1. test_cephfs_mirror_incremental_sync_with_type_mixup
```
2025-07-15T22:08:27.463+0000 7f13558ae640 10 client.8705 send_request client_request(unknown.0:89 readdir_snapdiff #0x10000000000 2025-07-15T22:08:27.468361+0000 caller_uid=0, caller_gid=0{}) v6 to mds.0
2025-07-15T22:08:27.463+0000 7f13558ae640  1 -- 172.21.15.151:0/3663776922 --> [v2:172.21.15.151:6834/1585733299,v1:172.21.15.151:6835/1585733299] -- client_request(unknown.0:89 readdir_snapdiff #0x10000000000 2025-07-15T22:08:27.468361+0000 caller_uid=0, caller_gid=0{}) v6 -- 0x5598a139e000 con 0x5598a6ba4c00
2025-07-15T22:08:27.463+0000 7f13558ae640 20 client.8705 awaiting reply|forward|kick on 0x7f13558a9ed0
2025-07-15T22:08:27.467+0000 7f13630c9640  1 -- 172.21.15.151:0/3663776922 <== mds.0 v2:172.21.15.151:6834/1585733299 101 ==== client_reply(???:89 = -22 (22) Invalid argument safe) v1 ==== 27+0+0 (crc 0 0 0) 0x5598621d7c00 con 0x5598a6ba4c00
2025-07-15T22:08:27.467+0000 7f13630c9640 20 client.8705 handle_client_reply got a reply. Safe:1 tid 89
2025-07-15T22:08:27.467+0000 7f13630c9640 10 client.8705 insert_trace from 2025-07-15T22:08:27.468444+0000 mds.0 is_target=0 is_dentry=0
2025-07-15T22:08:27.467+0000 7f13630c9640 10 client.8705 insert_trace -- no trace
2025-07-15T22:08:27.467+0000 7f13630c9640 20 client.8705 handle_client_reply signalling caller 0x7f13558a9ed0
2025-07-15T22:08:27.467+0000 7f13630c9640 20 client.8705 handle_client_reply awaiting kickback on tid 89 0x7f13630c59d0
2025-07-15T22:08:27.467+0000 7f13558ae640 20 client.8705 sendrecv kickback on tid 89 0x7f13630c59d0
2025-07-15T22:08:27.467+0000 7f13558ae640 20 client.8705 lat 0.001023
2025-07-15T22:08:27.467+0000 7f13558ae640 10 client.8705 _readdir_get_frag got error -22, setting end flag
2025-07-15T22:08:27.467+0000 7f13630c9640 20 client.8705 put_inode on 0x10000000000.snapdir(faked_ino=0 nref=38 ll_ref=0 cap_refs={} open={} mode=40775 size=0/0 nlink=1 btime=2025-07-15T22:07:43.961573+0000 mtime=2025-07-15T22:08:26.652492+0000 ctime=2025-07-15T22:08:26.652492+0000 change_attr=2 caps=- 0x5598a6c34a00\
) n = 1
2025-07-15T22:08:27.467+0000 7f13558ae640 -1 client.8705 readdir_snapdiff error: (22) Invalid argument
2025-07-15T22:08:27.467+0000 7f13558ae640  3 client.8705 closedir(0x5598621d6a80) = 0
2025-07-15T22:08:27.467+0000 7f13558ae640 10 client.8705 _closedir(0x5598621d6a80)
2025-07-15T22:08:27.467+0000 7f13558ae640 10 client.8705 _closedir detaching inode 0x559862217000
2025-07-15T22:08:27.467+0000 7f13558ae640 -1 cephfs::mirror::PeerReplayer(a5350d8d-9dd3-4c36-bd26-f305e2fcaa97) get_entry: failed to read directory=./file_z
2025-07-15T22:08:27.467+0000 7f13558ae640 20 cephfs::mirror::PeerReplayer(a5350d8d-9dd3-4c36-bd26-f305e2fcaa97) do_synchronize: r=-22
2025-07-15T22:08:27.467+0000 7f13558ae640 20 cephfs::mirror::PeerReplayer(a5350d8d-9dd3-4c36-bd26-f305e2fcaa97) finish_sync
2025-07-15T22:08:27.467+0000 7f13558ae640 20 cephfs::mirror::PeerReplayer(a5350d8d-9dd3-4c36-bd26-f305e2fcaa97) finish_sync: closing local directory=./file_z
2025-07-15T22:08:27.467+0000 7f13558ae640 20 client.8705 put_inode on 0x100000001fa.3(faked_ino=0 nref=12 ll_ref=0 cap_refs={} open={} mode=40775 size=0/0 nlink=1 btime=2025-07-15T22:08:26.434133+0000 mtime=2025-07-15T22:08:26.434133+0000 ctime=2025-07-15T22:08:26.434133+0000 change_attr=0 caps=pAsLsXs parents=0x1000\
0000000.3["file_z"] 0x559862217000) n = 1
```

```
2025-07-15T22:08:27.467+0000 7f13558ae640 20 cephfs::mirror::PeerReplayer(a5350d8d-9dd3-4c36-bd26-f305e2fcaa97) do_synchronize cur:10 prev:11 ret = -22
2025-07-15T22:08:27.467+0000 7f13558ae640 -1 cephfs::mirror::PeerReplayer(a5350d8d-9dd3-4c36-bd26-f305e2fcaa97) do_sync_snaps: failed to synchronize dir_root=/d0, snapshot=snap_1
2025-07-15T22:08:27.467+0000 7f13558ae640 -1 cephfs::mirror::PeerReplayer(a5350d8d-9dd3-4c36-bd26-f305e2fcaa97) sync_snaps: failed to sync snapshots for dir_root=/d0
2025-07-15T22:08:27.467+0000 7f13558ae640 20 cephfs::mirror::PeerReplayer(a5350d8d-9dd3-4c36-bd26-f305e2fcaa97) unregister_directory: dir_root=/d0
2025-07-15T22:08:27.467+0000 7f13558ae640 20 cephfs::mirror::PeerReplayer(a5350d8d-9dd3-4c36-bd26-f305e2fcaa97) unlock_directory: dir_root=/d0
2025-07-15T22:08:27.467+0000 7f13558ae640  3 client.8705 close enter(11)
```

2. test_cephfs_mirror_cancel_mirroring_and_readd

```
2025-07-15T21:21:21.143+0000 7f13550ad640  0 cephfs::mirror::PeerReplayer(747cb841-acb8-4f10-8cfc-da5072e9f38e) copy_to_remote: backing off r=-125
2025-07-15T21:21:21.143+0000 7f13550ad640 -1 cephfs::mirror::PeerReplayer(747cb841-acb8-4f10-8cfc-da5072e9f38e) remote_file_op: failed to copy path=./file.2: (125) Operation canceled
2025-07-15T21:21:21.143+0000 7f13550ad640 20 cephfs::mirror::PeerReplayer(747cb841-acb8-4f10-8cfc-da5072e9f38e) finish_sync
2025-07-15T21:21:21.143+0000 7f13550ad640 20 cephfs::mirror::PeerReplayer(747cb841-acb8-4f10-8cfc-da5072e9f38e) finish_sync: closing local directory=.
2025-07-15T21:21:21.143+0000 7f13550ad640 20 cephfs::mirror::PeerReplayer(747cb841-acb8-4f10-8cfc-da5072e9f38e) do_synchronize cur:10 prev:13 ret = -125
2025-07-15T21:21:21.143+0000 7f13550ad640 -1 cephfs::mirror::PeerReplayer(747cb841-acb8-4f10-8cfc-da5072e9f38e) do_sync_snaps: failed to synchronize dir_root=/d0, snapshot=snap0
2025-07-15T21:21:21.143+0000 7f13550ad640 -1 cephfs::mirror::PeerReplayer(747cb841-acb8-4f10-8cfc-da5072e9f38e) sync_snaps: failed to sync snapshots for dir_root=/d0
2025-07-15T21:21:21.143+0000 7f13550ad640 20 cephfs::mirror::PeerReplayer(747cb841-acb8-4f10-8cfc-da5072e9f38e) unregister_directory: dir_root=/d0
2025-07-15T21:21:21.143+0000 7f13550ad640 20 cephfs::mirror::PeerReplayer(747cb841-acb8-4f10-8cfc-da5072e9f38e) unlock_directory: dir_root=/d0
2025-07-15T21:21:21.147+0000 7f13550ad640 10 cephfs::mirror::PeerReplayer(747cb841-acb8-4f10-8cfc-da5072e9f38e) unlock_directory: dir_root=/d0 unlocked

```

vshankar commented:
@joscollin - as discussed, we need the additional commit from the main branch that was dropped in the latest update. And if there is a bug in that commit when backporting, then we should find that first.

vshankar commented:
> > @joscollin - as discussed, we need the additional commit from the main branch that was dropped in the latest update. And if there is a bug in that commit when backporting, then we should find that first.
> 
> I can debug and fix it. But as I mentioned in the discussion, that fix would make reef newer than main in git.

I don't understand how it will be _newer_ than main. The subclassing change is a commit that needs backporting since the main branch tests have been running with that.

> If we are backporting the entire file blockdiff feature, it makes sense. Otherwise, backporting subclassing alone doesn't fix anything in reef AFAIU.

The subclassing commit(*) has the remote and snapdiff sync class mechanism. The commit is standalone and does not reply on blockdiff.

(*): https://github.com/ceph/ceph/pull/62250/commits/d9ac4315154a73c7bc84e94f55a9966c115adaca#diff-902f4a8c220846756f7af155a115519a0382579cd04dcc7069121665c245edfe

joscollin commented:
jenkins test make check

joscollin commented:
jenkins test make check

joscollin commented:
> @joscollin any update on this?

@vshankar 
Yes, there is a bug in readdir_snapdiff. readdir_snapdiff->_readdir_get_frag fails (EINVAL) only for the directory type. The directory type delete --> recreate --> read fails. The interesting thing is it fails along with the fix https://github.com/ceph/ceph/commit/33c6f232069f1cb77447495657ab584e5d52f594.

I think @ifed01 also faced the same issue and working on a fix for it? I need to review this PR: https://github.com/ceph/ceph/pull/64995.

vshankar commented:
> > > > I think @ifed01 also faced the same issue and working on a fix for it? I need to review this PR: #64995.
> > > 
> > > 
> > > No, it doesn't fix the issue found in this PR. https://pulpito.ceph.com/jcollin-2025-08-19_15:07:28-fs:mirror-wip-jcollin-testing-reef-65222-1908_2-distro-default-smithi/8451654/
> > 
> > 
> > Yeh, I was about to leave a comment since I reviewed PR #64995 and it does not look like the test failed due to that bug.
> 
> I'm debugging this.

@joscollin - any update on this?

joscollin commented:
> > > > > I think @ifed01 also faced the same issue and working on a fix for it? I need to review this PR: #64995.
> > > > 
> > > > 
> > > > No, it doesn't fix the issue found in this PR. https://pulpito.ceph.com/jcollin-2025-08-19_15:07:28-fs:mirror-wip-jcollin-testing-reef-65222-1908_2-distro-default-smithi/8451654/
> > > 
> > > 
> > > Yeh, I was about to leave a comment since I reviewed PR #64995 and it does not look like the test failed due to that bug.
> > 
> > 
> > I'm debugging this.
> 
> @joscollin - any update on this?

@vshankar 
This backport has some failures, which could easily be avoided by dropping the commits from https://github.com/ceph/ceph/pull/62250/commits. Since the upcoming reef release is the final release, I think fixing it that way should be fine IMO.
